File: apple/PathGame/PathGame/Models/GameState.swift
//
//  GameState.swift
//  PathGame
//
//  Core game state model with Codable support for persistence
//

import Foundation

// MARK: - Position
struct Position: Codable, Equatable, Hashable {
    let row: Int
    let col: Int
    
    static func == (lhs: Position, rhs: Position) -> Bool {
        lhs.row == rhs.row && lhs.col == rhs.col
    }
}

// MARK: - Grid Size
enum GridSize: Int, Codable, CaseIterable {
    case small = 5
    case large = 7
    
    var n: Int {
        switch self {
        case .small: return 2
        case .large: return 3
        }
    }
    
    var displayName: String {
        "\(rawValue)×\(rawValue)"
    }
    
    var totalCells: Int {
        rawValue * rawValue
    }
}

// MARK: - Game State
struct GameState: Codable {
    var grid: [[Int]]
    var path: [Position]
    var currentPos: Position
    var gridSize: GridSize
    var gameOver: Bool
    var gaveUp: Bool
    var optimalLength: Int
    var optimalPath: [Position]
    var attempts: Int
    var dateString: String
    var consecutiveUndos: Int
    var usedBonusUndo: Bool
    
    var size: Int { gridSize.rawValue }
    var pathLength: Int { path.count }
    var isPerfect: Bool { pathLength == optimalLength && !gaveUp }
    
    // MARK: - Initialization
    init(gridSize: GridSize = .small, date: Date = Date()) {
        self.gridSize = gridSize
        self.grid = []
        self.path = []
        self.gameOver = false
        self.gaveUp = false
        self.optimalLength = 1
        self.optimalPath = []
        self.attempts = 1
        self.dateString = Self.dateToString(date)
        self.consecutiveUndos = 0
        self.usedBonusUndo = false
        
        // Generate grid
        let seed = Self.getDateSeed(from: date) + gridSize.n * 1000
        var rng = SeededRandom(seed: seed)
        
        let size = gridSize.rawValue
        for _ in 0..<size {
            var row: [Int] = []
            for _ in 0..<size {
                row.append(rng.nextInt(min: 1, max: 5))
            }
            self.grid.append(row)
        }
        
        // Start from center
        let center = gridSize.n
        self.currentPos = Position(row: center, col: center)
        self.path = [currentPos]
    }
    
    // MARK: - Date Helpers
    static func dateToString(_ date: Date) -> String {
        let calendar = Calendar.current
        let components = calendar.dateComponents([.year, .month, .day], from: date)
        return "\(components.year!)-\(components.month!)-\(components.day!)"
    }
    
    static func getDateSeed(from date: Date) -> Int {
        let dateStr = dateToString(date)
        var hash = 0
        for char in dateStr.unicodeScalars {
            let charValue = Int(char.value)
            hash = ((hash << 5) &- hash) &+ charValue
        }
        return abs(hash)
    }
    
    // MARK: - Move Validation
    func isValidMove(to position: Position) -> Bool {
        guard !gameOver else { return false }
        guard position.row >= 0, position.row < size,
              position.col >= 0, position.col < size else { return false }
        
        // Check if already visited
        guard !path.contains(position) else { return false }
        
        // Check if adjacent
        let dr = abs(position.row - currentPos.row)
        let dc = abs(position.col - currentPos.col)
        guard dr <= 1, dc <= 1, !(dr == 0 && dc == 0) else { return false }
        
        // Check value constraint
        let currentVal = grid[currentPos.row][currentPos.col]
        let targetVal = grid[position.row][position.col]
        guard abs(targetVal - currentVal) <= 1 else { return false }
        
        return true
    }
    
    func getValidMoves() -> [Position] {
        let directions = [
            (-1, -1), (-1, 0), (-1, 1),
            (0, -1),          (0, 1),
            (1, -1),  (1, 0),  (1, 1)
        ]
        
        return directions.compactMap { dr, dc in
            let pos = Position(row: currentPos.row + dr, col: currentPos.col + dc)
            return isValidMove(to: pos) ? pos : nil
        }
    }
    
    // MARK: - Move Execution
    mutating func makeMove(to position: Position) -> Bool {
        guard isValidMove(to: position) else { return false }
        
        consecutiveUndos = 0
        currentPos = position
        path.append(position)
        
        // Check if no more valid moves
        if getValidMoves().isEmpty {
            gameOver = true
        }
        
        return true
    }
    
    mutating func undo() -> Bool {
        guard path.count > 1, !gameOver else { return false }
        
        // Check undo limits
        if gridSize == .small {
            guard consecutiveUndos < 1 else { return false }
        } else {
            let maxUndos = usedBonusUndo ? 1 : 2
            guard consecutiveUndos < maxUndos else { return false }
            if consecutiveUndos == 1 {
                usedBonusUndo = true
            }
        }
        
        consecutiveUndos += 1
        path.removeLast()
        currentPos = path.last!
        
        return true
    }
    
    mutating func reset() {
        guard !gaveUp else { return }
        
        attempts += 1
        path = [Position(row: gridSize.n, col: gridSize.n)]
        currentPos = path[0]
        gameOver = false
        consecutiveUndos = 0
        usedBonusUndo = false
    }
    
    mutating func giveUp() {
        gaveUp = true
        gameOver = true
        path = optimalPath
        currentPos = optimalPath.last ?? currentPos
    }
    
    // MARK: - Percentage & Scoring
    var percentage: Int {
        guard optimalLength > 0 else { return 0 }
        return Int((Double(pathLength) / Double(optimalLength)) * 100)
    }
    
    var tieredResponse: String {
        let responses5x5: [Int: String] = [
            13: "Novice Navigator!",
            14: "Path Finder!",
            15: "Route Ranger!",
            16: "Trail Blazer!",
            17: "Way Maker!",
            18: "Journey Master!",
            19: "Expedition Expert!",
            20: "Odyssey Oracle!",
            21: "Path Perfection!"
        ]
        
        let responses7x7: [Int: String] = [
            25: "Beginner Pathfinder!",
            30: "Master Trailblazer!",
            35: "Cosmic Navigator!",
            40: "Omnipotent Pathfinder!",
            45: "Absolute Perfection!"
        ]
        
        if gridSize == .small {
            return responses5x5[pathLength] ?? (pathLength > 21 ? "Impossible Achievement!" : "Keep exploring!")
        } else {
            for (threshold, response) in responses7x7.sorted(by: { $0.key > $1.key }) {
                if pathLength >= threshold {
                    return response
                }
            }
            return "Keep pushing!"
        }
    }
}

// MARK: - Seeded Random
struct SeededRandom {
    var seed: Int
    
    mutating func next() -> Double {
        seed = (seed &* 1103515245 &+ 12345) & 0x7fffffff
        return Double(seed) / Double(0x7fffffff)
    }
    
    mutating func nextInt(min: Int, max: Int) -> Int {
        Int(next() * Double(max - min + 1)) + min
    }
}

File: android/PathGame/app/src/main/java/com/nagusamecs/pathgame/data/model/GameState.kt
package com.nagusamecs.pathgame.data.model

import java.time.LocalDate
import kotlin.math.abs

/**
 * Represents a position on the game grid
 */
data class Position(
    val row: Int,
    val col: Int
) {
    fun isAdjacentTo(other: Position): Boolean {
        val rowDiff = abs(row - other.row)
        val colDiff = abs(col - other.col)
        return rowDiff <= 1 && colDiff <= 1 && !(rowDiff == 0 && colDiff == 0)
    }
}

/**
 * Grid size options - n value matches web version (n=2 for 5x5, n=3 for 7x7)
 */
enum class GridSize(val size: Int, val n: Int, val displayName: String) {
    SMALL(5, 2, "5×5"),
    LARGE(7, 3, "7×7");
    
    val totalCells: Int get() = size * size
    val center: Int get() = n // Center is always n (2 for 5x5, 3 for 7x7)
}

/**
 * Seeded random number generator - MUST match web version exactly!
 * Web uses: seed = (seed * 1103515245 + 12345) & 0x7fffffff
 * Then: return seed / 0x7fffffff for next()
 * And: Math.floor(next() * (max - min + 1)) + min for nextInt
 */
class SeededRandom(seed: Long) {
    private var state = seed
    
    private fun next(): Double {
        state = (state * 1103515245L + 12345L) and 0x7FFFFFFFL
        return state.toDouble() / 0x7FFFFFFF.toDouble()
    }
    
    fun nextInt(min: Int, max: Int): Int {
        return (next() * (max - min + 1)).toInt() + min
    }
    
    companion object {
        /**
         * Generate seed from date - MUST match web version exactly!
         * Web uses: dateStr = "${year}-${month+1}-${day}" then hash
         */
        fun getDateSeed(date: LocalDate): Long {
            val dateStr = "${date.year}-${date.monthValue}-${date.dayOfMonth}"
            var hash = 0L
            for (char in dateStr) {
                val charValue = char.code.toLong()
                hash = ((hash shl 5) - hash) + charValue
                hash = hash and 0xFFFFFFFFL // Keep as 32-bit
            }
            return abs(hash)
        }
        
        fun forDate(date: LocalDate, gridSize: GridSize): SeededRandom {
            val baseSeed = getDateSeed(date)
            val seed = baseSeed + gridSize.n * 1000
            return SeededRandom(seed)
        }
    }
}

/**
 * Represents the complete state of a game
 */
data class GameState(
    val gridSize: GridSize = GridSize.SMALL,
    val grid: List<List<Int>> = emptyList(),
    val path: List<Position> = emptyList(),
    val isComplete: Boolean = false,
    val optimalLength: Int = 0,
    val optimalPath: List<Position> = emptyList(),
    val date: LocalDate = LocalDate.now(),
    val attempts: Int = 1,
    val gaveUp: Boolean = false,
    val consecutiveUndos: Int = 0,
    val usedBonusUndo: Boolean = false
) {
    val currentPosition: Position?
        get() = path.lastOrNull()
    
    val currentValue: Int?
        get() = currentPosition?.let { grid[it.row][it.col] }
    
    val pathLength: Int
        get() = path.size
    
    val percentage: Int
        get() = if (optimalLength > 0) (pathLength * 100) / optimalLength else 0
    
    val isPerfect: Boolean
        get() = pathLength >= optimalLength && optimalLength > 0 && !gaveUp
    
    /**
     * Generate a new grid with seeded random values
     * Values are 1-5 to match web version exactly
     */
    fun generateGrid(): GameState {
        val random = SeededRandom.forDate(date, gridSize)
        val newGrid = List(gridSize.size) { _ ->
            List(gridSize.size) { _ ->
                random.nextInt(1, 5) // 1-5 like web version
            }
        }
        
        val center = gridSize.center
        val startPosition = Position(center, center)
        
        return copy(
            grid = newGrid,
            path = listOf(startPosition),
            isComplete = false
        )
    }
    
    /**
     * Check if a move to the given position is valid
     */
    fun isValidMove(position: Position): Boolean {
        // Check bounds
        if (position.row < 0 || position.row >= gridSize.size ||
            position.col < 0 || position.col >= gridSize.size) {
            return false
        }
        
        // Check if already visited
        if (path.contains(position)) {
            return false
        }
        
        // Check adjacency
        val current = currentPosition ?: return false
        if (!current.isAdjacentTo(position)) {
            return false
        }
        
        // Check value constraint (within ±1)
        val currentVal = currentValue ?: return false
        val targetVal = grid[position.row][position.col]
        return kotlin.math.abs(targetVal - currentVal) <= 1
    }
    
    /**
     * Get all valid moves from current position
     */
    fun getValidMoves(): List<Position> {
        val current = currentPosition ?: return emptyList()
        val moves = mutableListOf<Position>()
        
        for (dRow in -1..1) {
            for (dCol in -1..1) {
                if (dRow == 0 && dCol == 0) continue
                val newPos = Position(current.row + dRow, current.col + dCol)
                if (isValidMove(newPos)) {
                    moves.add(newPos)
                }
            }
        }
        
        return moves
    }
    
    /**
     * Make a move to the given position
     */
    fun makeMove(position: Position): GameState {
        if (!isValidMove(position)) return this
        return copy(
            path = path + position,
            consecutiveUndos = 0  // Reset consecutive undos on move
        )
    }
    
    /**
     * Undo the last move - with limits matching web/iOS
     * 5x5: Max 1 consecutive undo
     * 7x7: Max 2 consecutive undos (then 1 after bonus used)
     */
    fun undoMove(): GameState {
        if (path.size <= 1) return this
        if (isComplete || gaveUp) return this
        
        // Check undo limits
        val maxUndos = when (gridSize) {
            GridSize.SMALL -> 1
            GridSize.LARGE -> if (usedBonusUndo) 1 else 2
        }
        
        if (consecutiveUndos >= maxUndos) return this
        
        val newUsedBonusUndo = if (gridSize == GridSize.LARGE && consecutiveUndos == 1) true else usedBonusUndo
        
        return copy(
            path = path.dropLast(1),
            consecutiveUndos = consecutiveUndos + 1,
            usedBonusUndo = newUsedBonusUndo
        )
    }
    
    /**
     * Check if undo is available
     */
    fun canUndo(): Boolean {
        if (path.size <= 1) return false
        if (isComplete || gaveUp) return false
        
        val maxUndos = when (gridSize) {
            GridSize.SMALL -> 1
            GridSize.LARGE -> if (usedBonusUndo) 1 else 2
        }
        
        return consecutiveUndos < maxUndos
    }
    
    /**
     * Reset the game to the starting position
     */
    fun restart(): GameState {
        if (gaveUp) return this  // Can't restart after giving up
        
        val center = gridSize.center
        return copy(
            path = listOf(Position(center, center)),
            isComplete = false,
            attempts = attempts + 1,
            consecutiveUndos = 0,
            usedBonusUndo = false
        )
    }
    
    /**
     * Give up and reveal the optimal path
     */
    fun giveUp(): GameState {
        return copy(
            gaveUp = true,
            isComplete = true,
            path = optimalPath.ifEmpty { path }
        )
    }
    
    /**
     * Check if the cell at position is part of the path
     */
    fun isInPath(position: Position): Boolean = path.contains(position)
    
    /**
     * Check if the cell at position is the current cell
     */
    fun isCurrent(position: Position): Boolean = currentPosition == position
    
    /**
     * Check if the cell at position is the start cell
     */
    fun isStart(position: Position): Boolean = path.firstOrNull() == position
    
    /**
     * Get the index of a position in the path (for drawing path lines)
     */
    fun pathIndex(position: Position): Int = path.indexOf(position)
}
